P2l visualisation and fixes

christian-rauch opened the pull request:
This PR:
- fixes for the Point2Line task map
- visualisation (RViz Marker) for point, line and their shortest distance

Before merging:
- Is there way to obtain the TF prefix `exotica/` from the API or is it constant?
- Is there a commons epsilon defined, that I can access in the task maps?

VladimirIvan commented on exotations/task_maps/task_map/src/Point2Line.cpp:
```diff
@@ -47,7 +47,7 @@ Eigen::Vector3d Point2Line::distance(const Eigen::Vector3d &point, const Eigen::
     // t = ((s-p)*(e-s)) / (|e-s|^2)    (* denotes the dot product)
 
```
This function does not return the point to line distance, it returns the vector from the closes point on the line towards the  query point. The name of the function does not reflect this.

VladimirIvan commented on exotations/task_maps/task_map/src/Point2Line.cpp:
```diff
@@ -85,15 +85,114 @@ void Point2Line::update(Eigen::VectorXdRefConst x, Eigen::VectorXdRef phi, Eigen
-            J(i, j) = -dv.dot(Eigen::Map<const Eigen::Vector3d>(Kinematics.J[i].getColumn(j).vel.data)) / dv.norm();
+            if (phi(i) > 1e-9)
+            {
+                J(i, j) = -(dv.normalized()).dot(Eigen::Map<const Eigen::Vector3d>(Kinematics.J[i].getColumn(j).vel.data));
```
This is not the correct Jacobian if all three points are attached different robot links!
You need to apply chain rule on the function used to calculate `phi`.

christian-rauch replied:
The start and end points of the line are defined in the same frame (the base frame, "Base"). And the `Kinematics` provides me with the link point in the base frame.

The `Distance` task map minimises the point distances the same way:
https://github.com/ipab-slmc/exotica/blob/e2bb512d40174ca2078f8eacbdef6439ca82730a/exotations/task_maps/task_map/src/Distance.cpp#L62

Am I missing something here?

VladimirIvan commented:
The TF prefix is currently constant.
It is very easy to create a parameter to change it though. It lives in the KinematicTree and a couple of TaskMaps use the hard coded prefix.

There is no common epsilon.

christian-rauch commented:
I rebased the test from the the `wxm-point2line-fixes` onto this. The test still fails if the line is a proper line (start!=end), and passes if the line is reduced to a point (start=end).

wxmerkt commented:
If the line is reduced to a point, you could use the Point2Point ``Distance`` task map - unless I misunderstand the use case.

christian-rauch commented:
Yes, it just demonstrates that the gradients are similar to the numerical gradients when reduced to a point-to-point distance. The point-to-line objective shows the expected behaviour in my use cases. I don't know yet why its gradients differ from the numerical gradients.

wxmerkt commented:
Thanks, I think we should turn the test off then, leave a comment and open an issue to document this to fix it.

VladimirIvan commented:
No, don't turn the test off. It means the gradient is off. This is a very basic task map with no discontinuities to handle so the analytical and finite differences gradients must agree.

You have two cost functions, one for a line and one for a line segment, but the Jacobian for both is the same. That can't be right.

wxmerkt commented on exotations/task_maps/task_map/init/Point2Line.in:
```diff
@@ -7,3 +7,4 @@ extend <exotica/TaskMap>
 
 Required Eigen::Vector3d EndPoint;  // point in 'Base' frame that defines the end of the line segment
 Optional bool infinite = true;      // true: interpret 'EndPoint' as the direction of an infinite line
```
I just noticed all the other Initialisers start with capital letters for parameters, can we make this one the same? 

If we get this rebased then on master and a few of the erroneous commits squashed I think it's ready for merging :-)

christian-rauch commented:
With this new gradients, the P2L Jacobians still fail the test with errors ~1e-5.
The part:
```C++
for (int j = 0; j < J.cols(); j++)
{
    J.middleRows<3>(i * 3).col(j) = Kinematics[0].J[i].data.topRows<3>().col(j).dot(line/line.squaredNorm()) * line - Kinematics[0].J[i].data.topRows<3>().col(j);
}
```
is derived from
![derivation_of_point_to_line_task_gradients](https://user-images.githubusercontent.com/8226248/46017326-1baebd80-c0cf-11e8-8203-e5239cf93e51.png)

VladimirIvan commented:
Please squash the commits. Otherwise looks good.

wxmerkt commented:
Further, please rename the initialiser variables to start with an upper case letter as pointed out in the commit above.

christian-rauch commented:
The initialisers have been renamed by [this commit](https://github.com/ipab-slmc/exotica/pull/339/commits/c1b20539a94d36755c8c07669cebc8b1372a3914).

Are you going to merge, even though the test will fail?

wxmerkt commented:
The test passes on CI. We will merge once squashed :-)

christian-rauch commented:
Yes, because the P2L test is basically a Point-to-Point test (`EndPoint==BaseOffset`).
However, if I comment out the original Point-to-Line test, it will fail with an error ~1e-5. That is why I assume it is not ready for merging, and why I haven't squashed yet.
If you want to merge despite the error, I am happy to squash.

christian-rauch commented:
Apparently, the End-effector orientation test in `test_maps` fails with the same magnitude of Jacobian error: `Jacobian error out of bounds: 1.14662e-05`. I therefore assume that this is not related to the P2L task.

Therefore: squashed and rebased on master.
